core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala

```diff
@@ -133,6 +140,86 @@ trait DataRepoBigQuerySupport {
+  }
+
+  /**
+   * generates the SQL and config to send to BigQuery for use in the queryEntities() method
+   * @param dataProject project containing the BQ data
+   * @param viewName dataset/snapshot to query in BQ
+   * @param entityType table to query in BQ
+   * @param entityQuery user-supplied query criteria
+   * @return the object to pass to BQ to execute a query
+   */
+  def queryConfigForQueryEntities(dataProject: String, viewName: String, entityType: String, entityQuery: EntityQuery): QueryJobConfiguration = {
+    // generate BQ SQL for this entity
+    val query = s"SELECT * FROM `${sanitizeSql(dataProject)}.${sanitizeSql(viewName)}.${sanitizeSql(entityType)}` " +
+      s"ORDER BY ${sanitizeSql(entityQuery.sortField)} ${SortDirections.toSql(entityQuery.sortDirection)} " +
+      s"LIMIT ${entityQuery.pageSize} " +
+      s"OFFSET ${translatePaginationOffset(entityQuery).toLong};"
+
+    // generate query config
+    QueryJobConfiguration.newBuilder(query)
+      .build
+  }
+
+  /**
+   * Remove illegal/undesired characters from a string so that it is safe to use inside
+   * a BigQuery SQL statement.
+   *
+   * BQ does not support bind parameters everywhere we want to use them.
+   * Per https://cloud.google.com/bigquery/docs/parameterized-queries,
+   * "Parameters cannot be used as substitutes for identifiers, column names,
+   * table names, or other parts of the query."
+   *
+   * Therefore we must sanitize any dynamic strings we want to use in those places.
+   *
+   * Google's doc on table naming (https://cloud.google.com/bigquery/docs/tables) says:
+   *   - Contain up to 1,024 characters
+   *   - Contain letters (upper or lower case), numbers, and underscores
+   * Google's doc on dataset naming (https://cloud.google.com/bigquery/docs/datasets) has the exact same requirements
+   * Google's doc on view naming (https://cloud.google.com/bigquery/docs/views) has the exact same requirements
+   * Google's doc on project naming (https://cloud.google.com/resource-manager/docs/creating-managing-projects) says:
+   *   - The project ID must be a unique string of 6 to 30 lowercase letters, digits, or hyphens. It must start with a letter, and cannot have a trailing hyphen.
+   * Google's doc on column naming (https://cloud.google.com/bigquery/docs/schemas) says:
+   *   - A column name must contain only letters (a-z, A-Z), numbers (0-9), or underscores (_), and it must start with a letter or underscore. The maximum column name length is 128 characters
+   *
+   * Given those naming requirements, this function allows ONLY: letters, numbers, underscores, hyphens. It does not
+   * impose any length restrictions.
+   *
+   * @param input the string to be sanitized
+   * @return the sanitized string
+   */
+  def sanitizeSql(input: String): String = {
+    if (input == null || input.isEmpty)
+      input
+    else
+      DataRepoBigQuerySupport.illegalBQChars.replaceAllIn(input, "")
+  }
+
```

Comment:
Please may I have lots of eyes on this sanitization function, plus its calling usages and test coverage?

Reply:
shouldn't we throw an exception instead of stripping?

Reply:
fair point, I waffled on this choice

Reply:
+1 for throwing an exception and letting the caller decide how to fix it instead of fixing it on their behalf. sanitize and it's usage looks good to me as well.

Reply:
sanitization looks good to me as well. I also vote for throwing an exception - stripping might end up with our code throwing an exception in a different place, (user submits bad table name -> table name is stripped -> new table name can't be found in BQ -> user receives "<stripped table name> table not found error" -> user is confused why the table name in the error is different from their input)

Reply:
@davidangb 

1. > It does not impose any length restrictions.Is 64 just an arbitrary length to print out to logs/output in case illegal chars are found? 

2. I think it may be a good idea to quote user input to avoid any issues with BigQuery reserved keywords. https://cloud.google.com/bigquery/docs/reference/standard-sql/lexical#reserved_keywords. 
E.g. to use ORDER BY `${validateSql(entityQuery.sortField)}` (it's a bit of an edge case).  Another case would be when a sort field would not match the column name in the db but 
I guess that would not be so much of a problem since it'd return an error I guess.  Maybe something like: (https://henrikwarne.com/2020/03/22/secure-by-design/)

```
private static String quote(final String identifier) {
        return BIGQUERY_QUOTE_CHAR + identifier + BIGQUERY_QUOTE_CHAR;
    }
```



3. In https://github.com/broadinstitute/rawls/blob/4482e3a1ba25969341b7f6f937d13f275082e99e/core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala#L15 
`[^a-zA-Z0-9\-_]` can be probably simplified to `[^\w\-]+`    (if you care :)) 

4. Where applicable (not the case on project IDs or table names), parameterized queries should be used to prevent SQLi. https://cloud.google.com/bigquery/docs/parameterized-queries

Reply:
> 1. Is 64 just an arbitrary length to print out to logs/output in case illegal chars are found?

Yes, exactly - if the input is invalid, we only output/log the beginning of the (sanitized) input, to prevent huge payloads affecting logs/output

> I think it may be a good idea to quote user input to avoid any issues with BigQuery reserved keywords.

Good point - I'll add a story to quote these inputs where they are used.

> Another case would be when a sort field would not match the column name in the db but
I guess that would not be so much of a problem since it'd return an error I guess.

Yes, this would return an error, we have business logic elsewhere to test for that.

> Where applicable (not the case on project IDs or table names), parameterized queries should be used to prevent SQLi. https://cloud.google.com/bigquery/docs/parameterized-queries

Wherever possible, we do use parameterized queries! This sanitization code is only used to validate user input in the places where BigQuery does not support parameters.